daemon/server/router/container/container_routes.go

```diff
@@ -874,6 +879,45 @@ func handleSysctlBC(
 	return warning, nil
 }
 
+// handlePortBindingsBC handles backward-compatibility for empty port bindings.
+//
+// Before Engine v29.0, an empty list of port bindings for a container port was
+// treated as if a PortBinding with an unspecified IP address and HostPort was
+// provided. The daemon was doing this backfilling on ContainerStart.
+//
+// Preserve this behavior for older API versions but emit a warning for API
+// v1.52 and drop that behavior for newer API versions.
+//
+// See https://github.com/moby/moby/pull/50710#discussion_r2315840899 for more
+// context.
+func handlePortBindingsBC(hostConfig *container.HostConfig, version string) string {
+	var warning string
+
+	for portProto, bindings := range hostConfig.PortBindings {
+		if len(bindings) > 0 {
+			continue
+		}
+		if versions.GreaterThan(version, "1.52") && len(bindings) == 0 {
+			// Starting with API 1.53, no backfilling is done. An empty slice
+			// of port bindings is treated as "no port bindings" by the daemon,
+			// but it still needs to backfill empty slices when loading the
+			// on-disk state for containers created by older versions of the
+			// Engine. Drop the PortBindings entry to ensure that no backfilling
+			// will happen when restarting the daemon.
+			delete(hostConfig.PortBindings, portProto)
+			continue
+		}
+
+		if versions.Equal(version, "1.52") {
+			warning = fmt.Sprintf("Container port %s has an empty list of port-bindings. Starting with API 1.53, this will be discarded.", portProto)
```

Comment:
If multiple port bindings trigger the warning, why only list the last one iterated upon?
```go
for ... {
        if versions.Equal(version, "1.52") {
                warningPorts = append(warningPorts, portProto)
        }
}
if len(warningPorts) > 0 {
        return fmt.Sprintf("One or more container ports has an empty list of port-bindings. Starting with API 1.53, the bindings for %s will be discarded.", strings.Join(warningPorts, ", "))
}
return ""
```